x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheDirectory.java

```diff
@@ -0,0 +1,156 @@
+        private final long end;
+
+        /** a releasable provided by the CacheService that must be released when this IndexInput is done with the cached file **/
+        private @Nullable Releasable releasable;
+        private List<IndexInput> clones;
+        private boolean closed;
+
+        CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext) {
+            this(fileName, fileLength, ioContext, "CachedBufferedIndexInput(" + fileName + ")", 0L, fileLength);
+        }
+
+        private CacheBufferedIndexInput(String fileName, long fileLength, IOContext ioContext, String desc, long offset, long length) {
+            super(desc, ioContext);
+            this.fileName = fileName;
+            this.fileLength = fileLength;
+            this.file = cacheDir.resolve(fileName);
+            this.ioContext = ioContext;
+            this.offset = offset;
+            this.end = offset + length;
+            this.clones = new ArrayList<>();
+        }
+
+        @Override
+        public long length() {
+            return end - offset;
+        }
+
+        @Override
+        protected void readInternal(final byte[] b, final int off, final int len) throws IOException {
+            final long pos = getFilePointer() + offset;
+            releasable =
+                cacheService.readFromCache(file, fileLength, releasable, () -> in.openInput(fileName, ioContext), pos, b, off, len);
+        }
+
+        @Override
+        protected void seekInternal(long pos) throws IOException {
+            if (pos > length()) {
+                throw new EOFException("Reading past end of file [position=" + pos + ", length=" + length() + "] for " + toString());
+            } else if (pos < 0L) {
+                throw new IOException("Seeking to negative position [" + pos + "] for " + toString());
+            }
+        }
+
+        @Override
+        public CacheBufferedIndexInput clone() {
+            final CacheBufferedIndexInput clone = (CacheBufferedIndexInput) super.clone();
+            clone.clones = new ArrayList<>();
+            clone.releasable = null;
+            clones.add(clone);
+            return clone;
+        }
+
+        @Override
+        public IndexInput slice(String sliceDescription, long offset, long length) throws IOException {
+            if (offset < 0 || length < 0 || offset + length > this.length()) {
+                throw new IllegalArgumentException("slice() " + sliceDescription + " out of bounds: offset=" + offset
+                    + ",length=" + length + ",fileLength=" + this.length() + ": " + this);
+            }
+            final CacheBufferedIndexInput slice = new CacheBufferedIndexInput(fileName, fileLength, ioContext,
+                getFullSliceDescription(sliceDescription), this.offset + offset, length);
+            clones.add(slice);
+            return slice;
+        }
+
+        @Override
+        public void close() {
+            if (closed == false) {
```

Comment:
Maybe use an `AtomicBoolean` for thread-safety, and then

```suggestion
            if (closed.compareAndSet(false, true)) {
```

Probably wise to `assert closed.get() == false` in the other methods too.